analyses/comparative-RNASeq-analysis/01-correlation-matrix.py

```diff
@@ -11,34 +12,84 @@
 import sys
 import argparse
 import math
+import tarfile
 import numpy as np
 import pandas as pd
 import scipy.stats
 import sklearn.metrics.pairwise as sklp
 import utils
 
-def prepare_expression(input_matrix,
-                       scratch_dir,
+def get_qc_status(mend_results_path, mend_manifest_path):
+    """Takes: mend results tgz containing bam_umend_qc.tsv files, and
+    manifest containing mapping from filename to sample id.
+    returns dictionary from sample id to mend status: PASS or FAIL"""
+    mend_tgz = tarfile.open(mend_results_path)
+    qc_files = (i for i in mend_tgz.getmembers() if i.name.endswith("bam_umend_qc.tsv"))
+    # Dictionary of filename (UUID.bam_umend_qc.tsv) to PASS or FAIL string
+    filename_map = { i.name : mend_tgz.extractfile(i).readlines()[1][-5:-1].decode("utf-8") for i in qc_files }
+    # Find sample ID entries in manifest and map them via filename.
+    manifest = utils.read_tsv(mend_manifest_path)
+    return { manifest.loc[k]["Kids.First.Biospecimen.ID"] : v for k, v in filename_map.items() }
+
+def filter_samples(input_matrix,
+                   mend_results_path,
+                   mend_manifest_path,
+                   clinical_path,
+                   nofilter=False,
+                   verbose=False):
+    """Filters samples by QC and tumor status.
+    Takes path to samples expression file; tgz of MEND QC results;
+    manifest of MEND QC files to sample ids; and clinical data file.
+    Returns data frame of sample expression filtered to contain only
+    MEND QC pass tumor samples. Filters are skippable via nofilter if,
+    for example, clinical and qc files aren't available."""
+    print_v = print if verbose else lambda *a, **k: None
+
+    print_v("Loading sample matrix {}".format(input_matrix))
+    raw_samples = utils.read_rds(input_matrix)
+
+    if nofilter:
+        print_v("Skipped the QC and tumor status filters; using all samples.")
+        return raw_samples
+
+    # Filter to only samples with a MEND QC status of PASS
+    qc_status = get_qc_status(mend_results_path, mend_manifest_path)
+    qc_pass_ids= [k for k, v in qc_status.items() if v == "PASS"]
+    qc_pass_samples = raw_samples[set(raw_samples.columns).intersection(qc_pass_ids)]
```

Comment:
While it should be the case that every sample has a QC call, is that something we ever check? I'd assume you would not want to include a sample without a matched QC, but it might be worth calling out separately in the log?

Reply:
Added a check for samples that didn't have matched QC. Since this indicates missing QC rather than existing QC that has somehow gone wrong (as your previous comment does), I'm comfortable moving forward without those samples rather than crashing the whole analysis.

Reply:
Yup, this seems like correct approach.